fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneSubColumnsRule.java

```diff
@@ -0,0 +1,319 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.rule.tree;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.starrocks.catalog.ColumnAccessPath;
+import com.starrocks.catalog.FunctionSet;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptExpressionVisitor;
+import com.starrocks.sql.optimizer.base.ColumnRefSet;
+import com.starrocks.sql.optimizer.operator.Operator;
+import com.starrocks.sql.optimizer.operator.OperatorBuilderFactory;
+import com.starrocks.sql.optimizer.operator.logical.LogicalProjectOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalScanOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CollectionElementOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperatorVisitor;
+import com.starrocks.sql.optimizer.operator.scalar.SubfieldOperator;
+import com.starrocks.sql.optimizer.rewrite.ReplaceColumnRefRewriter;
+import com.starrocks.sql.optimizer.task.TaskContext;
+import com.starrocks.thrift.TAccessPathType;
+
+import java.util.Deque;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class PruneSubColumnsRule implements TreeRewriteRule {
+
+    @Override
+    public OptExpression rewrite(OptExpression root, TaskContext taskContext) {
+        SubColumnAccessPathComputer computer = new SubColumnAccessPathComputer();
+        return root.getOp().accept(computer, root, null);
+    }
+
+    /*
+     * collect all complex expressions, such as: MAP_KEYS, MAP_VALUES, map['key'], struct.a.b.c ...
+     */
+    private static class ComplexExpressionCollector extends ScalarOperatorVisitor<Void, Void> {
+        private static final List<String> SUPPORT_FUNCTIONS = ImmutableList.<String>builder()
+                .add(FunctionSet.MAP_KEYS, FunctionSet.MAP_SIZE, FunctionSet.MAP_VALUES)
+                .build();
+
+        private final List<ScalarOperator> complexExpressions = Lists.newArrayList();
+
+        @Override
+        public Void visit(ScalarOperator scalarOperator, Void context) {
+            for (ScalarOperator child : scalarOperator.getChildren()) {
+                child.accept(this, context);
+            }
+            return null;
+        }
+
+        @Override
+        public Void visitVariableReference(ColumnRefOperator variable, Void context) {
+            complexExpressions.add(variable);
```

Comment:
Is every column need to be treated as a complexExpression?

Reply:
emmm, it's error, nice work